### ISSUE TITLE
Make datacontenttype value check case-insensitive

### ISSUE THREAD

**Issue by batrived** [cloudevents/sdk-csharp#282]
SDK is using case-sensitive checks when formatting CloudEvents based on "datacontenttype" value. MIME types type and subtype are case insensitive according to the spec. 

For example, this code snippet throws exception:
System.ArgumentException: 'JsonEventFormatter cannot serialize data of type System.String with content type 'application/JSON''

CloudEvent cloudEvent = new CloudEvent
{
    Id = "event-id",
    Type = "event-type",
    Source = new Uri("https://cloudevents.io/"),
    Time = DateTimeOffset.UtcNow,
    DataContentType = "application/**JSON**",
    Data = "text"
};

CloudEventFormatter formatter = new JsonEventFormatter();
var httpContent = cloudEvent.ToHttpContent(ContentMode.Structured, formatter);

**Comment by jskeet** [comment]
Could you clarify which spec (ideally with a link to the specific part) you mean? Is this a MIME types spec, or the CloudEvents spec? https://www.rfc-editor.org/rfc/rfc2046.html states that the charset parameter isn't case-sensitive, and that subtype names of image are not case-sensitive, but I don't see anything similar for "application".

**Comment by batrived** [comment]
@jskeet https://www.rfc-editor.org/rfc/rfc2045#section-2 

From the above rfc:

All media type values, subtype values, and
   parameter names as defined are case-insensitive.  However, parameter
   values are case-sensitive unless otherwise specified for the specific
   parameter.

**Comment by captainsafia** [comment]
We may want to take a look at the code ASP.NET uses for header comparisons ([ref](https://source.dot.net/#Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs,385)) here. It's used in ASP.NET to validate `Accepts` headers and handle the case-sensitivity rules correctly, AFAIK. It also has some capabilities to filter out invalid subtypes like th `application/cloudeventsfoobar` that is mentioned in [this TODO](https://github.com/cloudevents/sdk-csharp/blob/4fec77c9e9d5a1d907a469f0d059d22f861971e5/src/CloudNative.CloudEvents/Http/HttpClientExtensions.cs#L341).

**Comment by jskeet** [comment]
> We may want to take a look at the code ASP.NET uses for header comparisons ([ref](https://source.dot.net/#Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs,385)) here.

Yes, it's a shame that's not all in System.Net.Http.Headers.MediaTypeHeaderValue :( I suspect we'll need a really careful sweep through the code to find everything. I'll start keeping a log here...

**Comment by captainsafia** [comment]
> Yes, it's a shame that's not all in System.Net.Http.Headers.MediaTypeHeaderValue :( 

Yes, the `Microsoft.Net.Headers` stuff was specifically optimized for server/web scenarios as opposed to the "general purpose" types in `System.Net.Http.Headers`. We can get away with doing source-based code sharing to bring in the functionality from `Microsoft.Net.Headers` for this.
